lhotse/dataset/collation.py

```diff
@@ -8,19 +8,115 @@
 from lhotse.cut import MixedCut
 
 
-def collate_features(cuts: CutSet) -> torch.Tensor:
+class TokenCollater:
+    """Collate list of tokens
+
+    Map sentences to integers. Sentences are padded to equal length.
+    Beginning and end-of-sequence symbols can be added.
+    Call .inverse(tokens_batch, tokens_lens) to reconstruct batch as string sentences.
+
+    Example:
+        >>> token_collater = TokenCollater(cuts)
+        >>> tokens_batch, tokens_lens = token_collater(cuts.subset(first=32))
+        >>> original_sentences = token_collater.inverse(tokens_batch, tokens_lens)
+
+    Returns:
+        tokens_batch: IntTensor of shape (B, L)
+            B: batch dimensoion, number of input sentences
+            L: length of the longest sentence
+        tokens_lens: IntTensor of shape (B,)
+            Length of each sentence after adding <eos> and <bos>
+            but before padding.
+    """
+    def __init__(
+        self,
+        cuts: CutSet,
+        add_eos: bool = True,
+        add_bos: bool = True,
+        pad_symbol: str = "<pad>",
+        bos_symbol: str = "<bos>",
+        eos_symbol: str = "<eos>",
+        unk_symbol: str = "<unk>",
+    ):
+        assert all(len(cut.supervisions) == 1 for cut in cuts), \
+            "Expects only Cuts with a single supervision"
+
+        self.pad_symbol = pad_symbol
+        self.bos_symbol = bos_symbol
+        self.eos_symbol = eos_symbol
+        self.unk_symbol = unk_symbol
+
+        self.add_eos = add_eos
+        self.add_bos = add_bos
+
+        tokens = {
+            char for cut in cuts
+            for char in cut.supervisions[0].text
+        }
+        tokens_unique = (
+            [pad_symbol, unk_symbol]
+            + ([bos_symbol] if add_bos else [])
+            + ([eos_symbol] if add_eos else [])
+            + sorted(tokens)
+        )
+
+        self.token2idx = {token: idx for idx, token in enumerate(tokens_unique)}
+        self.idx2token = [token for token in tokens_unique]
```

Comment:
it is a cosmetic change but consider sorting idx2token and then creating token2idx from that; it might be a bit faster to debug/lookup stuff

Reply:
Thanks, good suggestion. Will adjust.

Reply:
Looking more into it, sorting the idx2token would mean that the pad symbol would not have the index zero, which is default padding index eg in pytorch Embedding layer. Or am I missing something?

Reply:
Oh sorry I didn't notice it is already sorted a few lines before. My bad!